core/templates/dev/head/pages/exploration-player-page/services/exploration-player-state.service.spec.ts

```diff
@@ -0,0 +1,142 @@
+  var UrlService = null;
+  var $rootScope = null;
+  var $q = null;
+
+  beforeEach(angular.mock.module('oppia'));
+  beforeEach(angular.mock.module('oppia', function($provide) {
+    var ugs = new UpgradedServices();
+    for (let [key, value] of Object.entries(ugs.getUpgradedServices())) {
+      $provide.value(key, value);
+    }
+  }));
+
+  beforeEach(function() {
+    angular.mock.module(function($provide) {
+      $provide.value('UrlService', {
+        getExplorationVersionFromUrl: function() {
+          return null;
+        },
+        getStoryIdInPlayer: function() {
+          return '1';
+        }
+      });
+
+      $provide.value('ContextService', {
+        isInExplorationEditorPage: function() {
+          return true;
+        },
+        isInQuestionPlayerMode: function() {
+          return false;
+        },
+        getExplorationId: function() {
+          return '123';
+        }
+      });
+      
+      $provide.constant('EXPLORATION_MODE', {
+        OTHER: false
+      });
+
+      $provide.value('AngularNameService', new AngularNameService());
+    });
+  });
+
+  beforeEach(angular.mock.inject(function(
+    _$q_, _$rootScope_, _ContextService_,
+    _EditableExplorationBackendApiService_,
+    _ExplorationFeaturesService_,
+    _ExplorationFeaturesBackendApiService_,
+    _ExplorationEngineService_,
+    _EXPLORATION_MODE_,
+    _ExplorationPlayerStateService_, _ExplorationStatesService_,
+    _NumberAttemptsService_,
+    _PlayerCorrectnessFeedbackEnabledService_,
+    _PlayerTranscriptService_,
+    _PlaythroughIssuesService_,
+    _PlaythroughService_,
+    _PretestQuestionBackendApiService_,
+    _QuestionBackendApiService_,
+    _QuestionPlayerEngineService_,
+    _ReadOnlyExplorationBackendApiService_,
+    _StateClassifierMappingService_,
+    _StatsReportingService_,
+    _UrlService_) {
+    $q = _$q_;
+    $rootScope = _$rootScope_;
+    ContextService = _ContextService_;
+    EditableExplorationBackendApiService = _EditableExplorationBackendApiService_;
+    ExplorationEngineService = _ExplorationEngineService_;
+    ExplorationFeaturesService = _ExplorationFeaturesService_;
+    ExplorationFeaturesBackendApiService = _ExplorationFeaturesBackendApiService_;
+    NumberAttemptsService = _NumberAttemptsService_;
+    PlayerCorrectnessFeedbackEnabledService = _PlayerCorrectnessFeedbackEnabledService_;
+    PlaythroughIssuesService = _PlaythroughIssuesService_;
+    PlayerTranscriptService = _PlayerTranscriptService_;
+    PlaythroughService = _PlaythroughService_;
+    PretestQuestionBackendApiService = _PretestQuestionBackendApiService_;
+    QuestionBackendApiService = _QuestionBackendApiService_;
+    QuestionPlayerEngineService = _QuestionPlayerEngineService_;
+    ReadOnlyExplorationBackendApiService = _ReadOnlyExplorationBackendApiService_;
+    StateClassifierMappingService = _StateClassifierMappingService_;
+    StatsReportingService = _StatsReportingService_;
+    UrlService = _UrlService_;
```

Comment:
Hi @kevintab95, I think the services being injected here further depend on some other services or constants and they need to be provided as well. I think instead of injecting, you can provide a mock for each of these to avoid injecting many services. Further services like `ContextService` & `UrlService` are already provided in `UpgradedServices.ts`, so you can avoid injecting them here.

Reply:
Thanks @ankita240796! I'll try mocking the required services.